src/Distributed/distributed_incompressible_model.jl

```diff
@@ -0,0 +1,97 @@
+using MPI
+
+using Oceananigans
+using Oceananigans.Grids
+using Oceananigans.Grids: halo_size
+
+#####
+##### Distributed model struct and constructor
+#####
+
+struct DistributedIncompressibleModel{A, G, M}
+    architecture :: A
+            grid :: G
+           model :: M
+end
+
+function DistributedIncompressibleModel(; architecture, grid, boundary_conditions=nothing, model_kwargs...)
+    my_rank = architecture.my_rank
+    i, j, k = architecture.my_index
+    Rx, Ry, Rz = architecture.ranks
+    my_connectivity = architecture.connectivity
+
+    ## Construct local grid
+
+    Nx, Ny, Nz = size(grid)
+
+    # Pull out left and right endpoints for full model.
+    xL, xR = grid.xF[1], grid.xF[Nx+1]
+    yL, yR = grid.yF[1], grid.yF[Ny+1]
+    zL, zR = grid.zF[1], grid.zF[Nz+1]
+    Lx, Ly, Lz = length(grid)
+
+    # Make sure we can put an integer number of grid points in each rank.
+    @assert isinteger(Nx / Rx)
+    @assert isinteger(Ny / Ry)
+    @assert isinteger(Nz / Rz)
+
+    nx, ny, nz = Nx÷Rx, Ny÷Ry, Nz÷Rz
+    lx, ly, lz = Lx/Rx, Ly/Ry, Lz/Rz
+
+    x₁, x₂ = xL + (i-1)*lx, xL + i*lx
+    y₁, y₂ = yL + (j-1)*ly, yL + j*ly
+    z₁, z₂ = zL + (k-1)*lz, zL + k*lz
+
+    # FIXME: local grid might have different topology!
+    my_grid = RegularRectilinearGrid(topology=topology(grid), size=(nx, ny, nz), x=(x₁, x₂), y=(y₁, y₂), z=(z₁, z₂), halo=halo_size(grid))
+
+    ## Change appropriate boundary conditions to halo communication BCs
+
+    # FIXME: Stop assuming (u, v, w, T, S).
```

Comment:
Would we want to do something like [this](https://github.com/CliMA/Oceananigans.jl/blob/0fb5286c9061c518f4b415a33df2f1ca39331323/src/TimeSteppers/runge_kutta_3.jl#L142) and first find the fields using `model_fields=fields(model)` and then loop over the fields?

Reply:
It occurs to me that we could store the BC names in a merged list that is ordered like fields

Reply:
Yes we should probably fix this before merging.

I think the complexity comes in mapping e.g. `u` to `UVelocityBoundaryConditions` but I'm sure it could be done without much work.

Reply:
We don't have to fix it before necessarily, and am happy to save this for letter if you prefer.  You've already done a lot here.